Remove as_tensor epilogue from float outputs

bobrenjc93 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #151885
* #151766

Fixes #151470

This was originally added as a nicety - ensuring all graphs maintain
the invariant that ouputs are tensors. This, however, breaks simple
graphs such as

```
        def f(y):
            y_scalar = y.item()
            return y_scalar
```

since inductor only knows how to codegen tensors and sizevars (aka
ints), it ends up barfing when doing the codegen of this epilogue.
Notice in particular for the following example, it doesn't correctly
codegen the float kernel argument.

```
cpp_fused_scalar_tensor_0 = async_compile.cpp_pybinding(['double*'], '''
extern "C"  void kernel(double* out_ptr0)
{
    {
        {
            {
                auto tmp0 = zuf0;
                auto tmp1 = c10::convert<double>(tmp0);
                out_ptr0[static_cast<int64_t>(0L)] = tmp1;
            }
        }
    }
}
''')
```

Since this is just a nicety, let's get rid of it for now until we
officially support floats in inductor. By doing this we stay in
python when doing item() calls, resulting in same behavior for eager vs
compile

```
def call(args):
    arg0_1, = args
    args.clear()
    assert_size_stride(arg0_1, (), ())
    zuf0 = arg0_1.item()
    buf0 = None
    del arg0_1
    return (zuf0, )
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

bobrenjc93 committed:
Remove as_tensor epilogue from float outputs

This was originally added as a nicety - ensuring all graphs maintain
the invariant that ouputs are tensors. This, however, breaks simple
graphs such as

```
        def f(y):
            y_scalar = y.item()
            return y_scalar
```

since inductor only knows how to codegen tensors and sizevars (aka
ints), it ends up barfing when doing the codegen of this epilogue.
Notice in particular for the following example, it doesn't correctly
codegen the float kernel argument.

```
cpp_fused_scalar_tensor_0 = async_compile.cpp_pybinding(['double*'], '''
extern "C"  void kernel(double* out_ptr0)
{
    {
        {
            {
                auto tmp0 = zuf0;
                auto tmp1 = c10::convert<double>(tmp0);
                out_ptr0[static_cast<int64_t>(0L)] = tmp1;
            }
        }
    }
}
''')
```

Since this is just a nicety, let's get rid of it for now until we
officially support floats in inductor. By doing this we stay in
python when doing item() calls, resulting in same behavior for eager vs
compile

```
def call(args):
    arg0_1, = args
    args.clear()
    assert_size_stride(arg0_1, (), ())
    zuf0 = arg0_1.item()
    buf0 = None
    del arg0_1
    return (zuf0, )
```

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Remove as_tensor epilogue from float outputs"


This was originally added as a nicety - ensuring all graphs maintain
the invariant that ouputs are tensors. This, however, breaks simple
graphs such as

```
        def f(y):
            y_scalar = y.item()
            return y_scalar
```

since inductor only knows how to codegen tensors and sizevars (aka
ints), it ends up barfing when doing the codegen of this epilogue.
Notice in particular for the following example, it doesn't correctly
codegen the float kernel argument.

```
cpp_fused_scalar_tensor_0 = async_compile.cpp_pybinding(['double*'], '''
extern "C"  void kernel(double* out_ptr0)
{
    {
        {
            {
                auto tmp0 = zuf0;
                auto tmp1 = c10::convert<double>(tmp0);
                out_ptr0[static_cast<int64_t>(0L)] = tmp1;
            }
        }
    }
}
''')
```

Since this is just a nicety, let's get rid of it for now until we
officially support floats in inductor. By doing this we stay in
python when doing item() calls, resulting in same behavior for eager vs
compile

```
def call(args):
    arg0_1, = args
    args.clear()
    assert_size_stride(arg0_1, (), ())
    zuf0 = arg0_1.item()
    buf0 = None
    del arg0_1
    return (zuf0, )
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

eellison reviewed (commented):
Not sure if this has implications for cudagraphs.. Maybe we should wait till cudagraph partitioning is rolled out ?

Do you know what the original motivation for this was ?

BoyuanFeng commented:
Major motivation is to fix a `KeyError: zuf0` when compiling the following code. (#151470)

```python
import torch

torch._dynamo.config.capture_scalar_outputs = True

def f(x, y):
    x1 = x + 1
    y_scalar = y.item()
    z = x1 + y_scalar
    return z, y_scalar


f = torch.compile(f)

f(torch.randn(2,3, device='cuda'), torch.tensor(3.0, device='cuda'))
```

I have not seen it in real model yet.

bobrenjc93 committed:
Update on "Remove as_tensor epilogue from float outputs"


Fixes #151470

This was originally added as a nicety - ensuring all graphs maintain
the invariant that ouputs are tensors. This, however, breaks simple
graphs such as

```
        def f(y):
            y_scalar = y.item()
            return y_scalar
```

since inductor only knows how to codegen tensors and sizevars (aka
ints), it ends up barfing when doing the codegen of this epilogue.
Notice in particular for the following example, it doesn't correctly
codegen the float kernel argument.

```
cpp_fused_scalar_tensor_0 = async_compile.cpp_pybinding(['double*'], '''
extern "C"  void kernel(double* out_ptr0)
{
    {
        {
            {
                auto tmp0 = zuf0;
                auto tmp1 = c10::convert<double>(tmp0);
                out_ptr0[static_cast<int64_t>(0L)] = tmp1;
            }
        }
    }
}
''')
```

Since this is just a nicety, let's get rid of it for now until we
officially support floats in inductor. By doing this we stay in
python when doing item() calls, resulting in same behavior for eager vs
compile

```
def call(args):
    arg0_1, = args
    args.clear()
    assert_size_stride(arg0_1, (), ())
    zuf0 = arg0_1.item()
    buf0 = None
    del arg0_1
    return (zuf0, )
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]
